### ISSUE TITLE
No support for GCC in Windows

### ISSUE THREAD

**Issue by TimSimpson** [lasote/conan-boost#13]
There is no support for GCC in Windows. To make matters worse, no good error message really states this- the error seen is 

```
conan install .. -s "compiler=gcc"  -s compiler.version=6.1 -s compiler.libcxx=libstdc++11 -s build_type=Debug -s arch=x86_64
...
ERROR: Error in Boost/1.60.0@lasote/stable
        package_info()
        invalid literal for int() with base 10: '6.1'
```

This is because a branch of Python code decides if the OS is Windows only Visual Studio should be considered, and it tries to convert compiler.version to an integer.

At the very least a better error message should be added. However - and I know this would probably be very difficult - it would be great if support would be added for GCC because right now it is blocking me from using MinGW in Windows on all of my projects (and this is right after I've been telling everyone Conan is the messianic build tool to save C++, with an emphasis on how easy it was to include Boost as a dependency).

**Comment by lasote** [comment]
Hi Tim, Thanks for notice it. We are currently adding some features to conan and conan_package_tools for ease the integration with MinGW and generate packages for MinGW. But it won't be available until conan 0.11
Would be great if could you fork this repository and "repair" the recipe to support MinGW, I will continue the work to auto-generate all the packages when 0.11 is out. Could you do it?:

1. Fork the repository
2. edit the conanfile.py 
3. execute "conan export lasote/stable" or export to your namespace
4. execute "conan test" to see if it works with MinGW.

If you have any problem in the process I can help you but I don't know if I have enough time to do this quick.

Thanks!

**Comment by lasote** [comment]
Hi, I have some free time this evening. Will report with advances.

**Comment by lasote** [comment]
Try now! binaries are not available, but update the recipe and build from sources. It should work:
`conan install Boost/1.60.0@lasote/stable -u --build` or remove the recipe and install your requires again `conan remove Boost/1.60.0@lasote/stable` and `conan install --build missing`.

With conan 0.11 we will introduce 2 new settings for the Mingw compiler (thread and exception), also we are working in conan packages for applications, a Mingw installer will be available with a very useful "virtual environment' that will allow us to change between mingw version just with a command (4.8, 4.9, with exception seh, dwarf2, sjlj, posix threads or win32... etc).

So you can keep saying that conan is the messianic build tool to save C++ ;) It's just a matter of improve a little the conan package recipes, conan is very generic but the recipes requires some community collaboration like you! :+1:

**Comment by TimSimpson** [comment]
Nice! Thanks for your hard work.

For now I attempted to build it and it hit a build error. I'm using STL's version of MinGW and its possible my environment is a little weird. I'm a bit busy tonight but I'll try to look into this soon as I'd love to try to help out here.

**Comment by lasote** [comment]
Ok! If you could paste here the error and your settings etc I can help you

**Comment by TimSimpson** [comment]
Hi @lasote . Sorry for taking so long to get to this but I've been pretty busy.

Here's a relevant bit of stack trace:
```
common.mkdir bin.v2\libs\date_time\build\gcc-mngw-6.1.0\dbg\lnk-sttc\thrd-mlt
common.mkdir bin.v2\libs\date_time\build\gcc-mngw-6.1.0\dbg\lnk-sttc\thrd-mlt\gregorian
gcc.compile.c++ bin.v2\libs\context\build\gcc-mngw-6.1.0\dbg\lnk-sttc\thrd-mlt\windows\stack_traits.o
libs\context\src\windows\stack_traits.cpp: In function 'SYSTEM_INFO {anonymous}::system_info()':
libs\context\src\windows\stack_traits.cpp:66:17: error: 'once_flag' in namespace 'std' does not name a type
     static std::once_flag flag;
                 ^~~~~~~~~
libs\context\src\windows\stack_traits.cpp:67:5: error: 'call_once' is not a member of 'std'
     std::call_once( flag, [](){ ::GetSystemInfo( & si); } );
     ^~~
libs\context\src\windows\stack_traits.cpp:67:21: error: 'flag' was not declared in this scope
     std::call_once( flag, [](){ ::GetSystemInfo( & si); } );
                     ^~~~
```
So it appears that Boost is not being compiled with C++11 support switched on. In my experience building Boost for MinGW I've needed to pass "cxxflags=-std=c++11" or something similar to it but since I'm using a version of MinGW based on GCC 6.0 I had thought this wouldn't be necessary.

I modified the conanfile to pass "-std=c++11" to b2 but had the same result.

I'm using Mingw from here btw: 
https://nuwen.net/mingw.html

There seems to be several competing flavors of MinGW these days. I'm not sure which one is more popular but if you know of one I should try instead I'll switch.

**Comment by lasote** [comment]
Could you provide me a simple example program that generates the error?

About the more popular flavor, in the next conan release we will support two new settings for MinGW (to handle exception system and threads model) and will provide a mechanism to install MinGW. We download the different flavors from: http://downloads.sourceforge.net/project/mingw-w64

**Comment by TimSimpson** [comment]
Hi @lasote . Sorry for being radio silent for so long but I just haven't had time to look into this lately.

First off - thanks for creating this and the phenomenal MinGW packages for Conan's new virtual env feature. It definitely makes  having an agreed upon recreate easier.

Anyway, here is the recreate:
https://github.com/TimSimpson/conan-boost-mingw-recreate

It makes a Conan virtualenv with MinGW 4.9 and then uses Boost which can't link when it needs to run a unit test, which is the same thing I'm now seeing on my other projects.

Here's a sampling of the errors:

```
D:/Tools/conan-data/mingw_installer/0.1/lasote/testing/package/e3fafa852fbf3267e0e3fc81c4f41a7360e26aef/bin/../lib/gcc/x86_64-w64-mingw32/4.9.2/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lboost_prg_exec_monitor
D:/Tools/conan-data/mingw_installer/0.1/lasote/testing/package/e3fafa852fbf3267e0e3fc81c4f41a7360e26aef/bin/../lib/gcc/x86_64-w64-mingw32/4.9.2/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lboost_test_exec_monitor
D:/Tools/conan-data/mingw_installer/0.1/lasote/testing/package/e3fafa852fbf3267e0e3fc81c4f41a7360e26aef/bin/../lib/gcc/x86_64-w64-mingw32/4.9.2/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lboost_atomic
D:/Tools/conan-data/mingw_installer/0.1/lasote/testing/package/e3fafa852fbf3267e0e3fc81c4f41a7360e26aef/bin/../lib/gcc/x86_64-w64-mingw32/4.9.2/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lboost_container
D:/Tools/conan-data/mingw_installer/0.1/lasote/testing/package/e3fafa852fbf3267e0e3fc81c4f41a7360e26aef/bin/../lib/gcc/x86_64-w64-mingw32/4.9.2/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lboost_date_time
D:/Tools/conan-data/mingw_installer/0.1/lasote/testing/package/e3fafa852fbf3267e0e3fc81c4f41a7360e26aef/bin/../lib/gcc/x86_64-w64-mingw32/4.9.2/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lboost_exception
```

Note there is a hack that makes almost all of the linker errors go away. If I enter the conan data directory and find the package for Boost 1.60 with MinGW, then copy all of the libraries so they don't include the variant details added by Boost Build (i.e. change libboost_atomic-mgw49-mt-1_60.a to libboost_atomic.a, libboost_unit_test_framework-mgw49-mt-1_60.a to libboost_unit_test_framework.a etc) then all of the Boost linker errors go away. After that all that remains is the zlib error. Here's the output:

```
[ 75%] Linking CXX executable bin\test_acme_exe.exe
D:/Tools/conan-data/mingw_installer/0.1/lasote/testing/package/e3fafa852fbf3267e0e3fc81c4f41a7360e26aef/bin/../lib/gcc/x86_64-w64-mingw32/4.9.2/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lzlibstatic
collect2.exe: error: ld returned 1 exit status
CMakeFiles\test_acme_exe.dir\build.make:97: recipe for target 'bin/test_acme_exe.exe' failed
mingw32-make.exe[2]: *** [bin/test_acme_exe.exe] Error 1
CMakeFiles\Makefile2:135: recipe for target 'CMakeFiles/test_acme_exe.dir/all' failed
mingw32-make.exe[1]: *** [CMakeFiles/test_acme_exe.dir/all] Error 2
makefile:82: recipe for target 'all' failed
mingw32-make.exe: *** [all] Error 2
ERROR: Error 2 while executing cmake --build .
```

**Comment by sixten-hilborn** [comment]
Hi @TimSimpson @lasote, the first errors that you solved with a "hack" is a duplicate of issue #15, for which I've created a PR (#41).
The second issue with zlib looks like an issue in the zlib package that has been fixed now.
